### ISSUE TITLE
[question] Using imgui, how to use imgui_stdlib.* files?

### ISSUE THREAD

**Issue by yudi-matsuzake** [conan-io/conan-center-index#13641]
I am using Conan 1.52.0 on linux and the package `imgui/cci.20220621+1.88.docking`. In the ImGui context, I want to use the code from `ocornut/imgui/misc/cpp/imgui_stdlib.*`. In the [recipe file](https://github.com/conan-io/conan-center-index/blob/master/recipes/imgui/all/conanfile.py) only the `backends/imgui_impl_*` are being copied to the `res` folder. Apparently, none of the `imgui/misc` files are being added to the package. Is there some way that I could use the `[imports]` section on these files?

**Comment by AndreyMlashkin** [comment]
@jgsogo @vbieleny @SpaceIm
ping recipe maintainers

**Comment by uilianries** [comment]
@AndreyMlashkin please read https://github.com/conan-io/conan-center-index/pull/13461

>  Apparently, none of the imgui/misc files are being added to the package.

@yudi-matsuzake What is the extension of `imgui_stdlib.*` ? Usually `res` folder is associated to configuration files which are needed by binaries or even libraries. What exactly do you need, headers, cpp files?

**Comment by yudi-matsuzake** [comment]
> @yudi-matsuzake What is the extension of `imgui_stdlib.*` ? Usually `res` folder is associated to configuration files which are needed by binaries or even libraries. What exactly do you need, headers, cpp files?

The files that I need are `imgui_stdlib.h` and `imgui_stdlib.cpp`; they are found in the `imgui/misc/cpp` folder and contains very simple `InputText()` wrappers for c++ standard library.

Indeed, maybe the `res` folder is not the place for these files and could be a more appropriate folder. Could be a good idea to bring the whole `misc` folder to the conan package? The `imgui/misc` folder contains [a lot of useful code](https://github.com/ocornut/imgui/blob/master/misc/README.txt) and in the future could be even more. There is some experimentation functions that could be merged in the ImGui's master branch like `imgui_scoped.{h,cpp}`.

**Comment by AndreyMlashkin** [comment]
Yes, totally agree with this approach.

For now there are only few libs using imgui and diligent-tools is one of them (I a maintainer)
It would simplify the way it handles the dependency, but most probably the change itself will break the existing recipe

**Comment by uilianries** [comment]
@yudi-matsuzake Could it be in `include/` instead? Otherwise, we will need to add a rule to Hooks. Anyway, it seems be a legit addition to the package.

**Comment by yudi-matsuzake** [comment]
> @yudi-matsuzake Could it be in `include/` instead? Otherwise, we will need to add a rule to Hooks. Anyway, it seems be a legit addition to the package.

Yes, I do think the `.h` files in `include/` folder would be appropriate but what about the `.cpp` files? Maybe you have a different approach in mind but if they are compiled and put in the same library I think people will pay for it (with library size bloat) even if they do not use the c++ standard library. Therefore, I don't know if this approach is a good idea.

I have zero experience with Conan package maintenance, and I don't have the knowledge of the best practices for this case; but could this be archived with some option argument? In this option, we could specify which additional `.cpp` files we could add to the library.

Another approach would be compiling each folder in `imgui/misc/` as a new library.

**Comment by jcar87** [comment]
From what I can see in the ImGui [readme](https://github.com/ocornut/imgui#usage):

> The core of Dear ImGui is self-contained within a few platform-agnostic files which you can easily compile in your application/engine. **They are all the files in the root folder of the repository (imgui*.cpp, imgui*.h).**

This matches what we build in the Conan package: 
https://github.com/conan-io/conan-center-index/blob/master/recipes/imgui/all/CMakeLists.txt#L14-L15

Even if we packaged the header files, you would still need the .cpp files to be compiled. These are in the `misc` folder and referred to as "examples":
https://github.com/ocornut/imgui/tree/master/misc/cpp

However the functions are still defined in the `ImGui` namespace with the `IMGUI_API` macro.

Imgui doesn't have "official" build scripts so it's really at the discretion of the users which bits to bring in and compile - we just bring the core files as per their documentation. 

From a usability perspective, we would err on the side of caution and not build "experimental" APIs into the libraries we provide. On the other hand these files have changed little since they were introduced - perhaps it's worth asking the ImGui maintainers whether they can be considered stable enough to be part of a distribution.

**Comment by yudi-matsuzake** [comment]
> Even if we packaged the header files, you would still need the .cpp files to be compiled. The files in question are in the `misc` folder and referred to as "examples": https://github.com/ocornut/imgui/tree/master/misc/cpp

Okay, I think I understand better now. Correct me if I am wrong but the correct approach would be to manually copy these files to my project, right? Is there a procedure using Conan to copy source files even if they are not in the package? If there is a way to `[imports]` those files, I believe there would be no need to packaged the compiled `*.cpp` files. Users could decide which bits to import, ideally using only Conan.

> From a usability perspective, we would err on the side of caution and not build "experimental" APIs into the libraries we provide. On the other hand these files have changed little since they were introduced - perhaps it's worth asking the ImGui maintainers whether they can be considered stable enough to be part of a distribution.

As far as I know, the `imgui/misc/cpp/imgui_stdlib.{h,cpp}` files are stable because there are some unstable features like `imgui_scoped.{h,cpp}` that is [not merged to the main repository](https://github.com/ocornut/imgui/blob/master/misc/cpp/README.txt):

> imgui_scoped.h
  [Experimental, not currently in main repository]

Apparently, the unstable features are not merged to the main branch. But of course, the direct communication with the ImGui developers would be more appropriate.

**Comment by samuel-emrys** [comment]
This sounds like it might be in scope for a second package, imgui-misc, perhaps?
